text/0000-refs.md

```diff
@@ -0,0 +1,209 @@
+<template>
+    <div lwc:ref="foo">Hello world!</div>
+</template>
+```
+
+```js
+// component.js
+import { LightningElement } from 'lwc';
+export default class extends LightningElement {
+  renderedCallback() {
+    console.log(this.refs.foo.textContent); // 'Hello world!'
+  }
+}
+```
+
+## Motivation
+
+As of today, the only way to access DOM nodes defined in the template via JavaScript is using standard DOM APIs on `this.template` (shadow DOM) or `this` (light DOM):
+
+```js
+// shadow DOM
+this.template.querySelector('div');
+
+// light DOM
+this.querySelector('div');
+```
+
+This works, and it follows vanilla web component conventions, but it's cumbersome. It forces the developer to think of a query selector (or another technique, like `.firstChild`) to locate an element in the DOM.
+
+In the context of light DOM components, there is an additional difficulty: `this.querySelector()` may return elements from outside of the immediate template (e.g. light DOM children). So developers have to choose their selectors carefully.
+
+### Prior art
+
+The concept of "refs" exists in many other frameworks:
+
+- [React: `createRef()`](https://reactjs.org/docs/refs-and-the-dom.html)
+- [Vue: template refs](https://v3.vuejs.org/guide/composition-api-template-refs.html)
+- [Svelte: `bind:this`](https://svelte.dev/docs#template-syntax-element-directives-bind-this)
+- [Lit: `createRef()`](https://lit.dev/docs/api/directives/#ref)
+- [Angular: `elementRef`](https://angular.io/api/core/ElementRef)
+- [Stencil: `ref`](https://stenciljs.com/docs/templating-jsx#getting-a-reference-to-a-dom-element)
+
+By building on these conventions, we offer a simple and ergonomic way to solve the problem of locating elements at runtime.
+
+## Detailed design
+
+The basic design of refs is the use of the `lwc:ref` directive in templates:
+
+```html
+<template>
+    <div lwc:ref="foo"></div>
+</template>
+```
+
+...which can be accessed at runtime in a component using `this.refs`:
+
+```js
+export default class extends LightningElement {
+  renderedCallback() {
+    console.log(this.refs.foo); // `<div>` DOM node
+  }
+}
+```
+
+### `lwc:ref`
+
+`lwc:ref` can only refer to a static string. For instance, this will throw a template compiler error:
+
+```html
+<template>
+    <div lwc:ref={dynamic}></div> <!-- Not allowed! -->
+</template>
+```
+
+`lwc:ref` cannot be applied to `<template>`s, or to `<slot>`s in light DOM:
+
+```html
+<template>
+    <template if:true={foo} lwc:ref="foo"></template> <!-- Not allowed! -->
+</template>
+```
+
+```html
+<template lwc:render-mode="light">
+    <slot lwc:ref="foo"></slot> <!-- Not allowed! -->
+</template>
+```
+
+### `this.refs`
+
+`this.refs` is a plain object, with string keys and DOM node values. If a ref does not exist, then `this.refs["unknown"]` returns `undefined`.
+
+The object is read-only, and attempting to add, modify, or delete its properties from within the component will result in a runtime error.

```

Comment:
So, I have some concerns about these two paragraphs. I don't know yet how do you plan to implement this, but here are my questions:

1. what is the nature of `this.refs` property? is it a configurable accessor?
2. `this.refs === this.refs`?
3. `this.refs === this.oldRefs` assuming `oldRefs` was cached in the previous call to `renderedCallback`? If the `refs` object is reused in between render calls, that will not work, because you you will have to mark it as non-extensible, meaning switching templates will break the object invariants.
4. what will `Reflect.ownKeys(this.refs)` return? is it going to be all `lwc:ref` names of the current selected template? or only those rendered in the DOM during the last rendering cycle?

Reply:
Here is my recommendation:

* `LightningElement.prototype.refs` property to be a writable and configurable data property. (this will be easier for backward compatibility.
* the `refs` object value to be defined as non-extensible, with non-configurable, non-writable data entries, and it is produced after each rendering cycle. I'm not 100% sure that this is the right approach though! Maybe a throw away object that is mutable, and its entries are configurable is better, thinking about inheritance is the problem here to preserve some kind of integrity.

Reply:
Regarding @caridy 's 4th point, I am wondering how does the `refs` object evolve throughout the lifecycle of the component:
- Shall all the know refs in the template be set to `undefined` in advance?
- When an element with a ref is removed, is the ref set to `undefined`, or is the property removed from the `refs` object?
- When the template is swapped is the previous `refs` object reused for the new template?

What does the `console.log(this.refs)` print the 3 times it is invoked in the following example?

```js
class Example extends LightningElement {
  visible = false;

  connectedCallback() {
    setTimeout(() => {
      this.visible = true;
      
      setTimeout(() => {
        this.visible = false;
      });
    })
  }

  renderedCallback() {
    console.log(this.refs); // Invoked 3 times
  }
}
```

```html
<template>
  <template if:true={visible}>
    <div lwc:ref="my-div"></div>
  </template>
</template>
```

Reply:
These are interesting questions. However we decide to represent the object will have an impact on perf and usability.

For instance: if we recreate the `this.refs` object on every render, then that's potentially bad for perf because we're creating and destroying lots of objects. Plus, the user may cache `this.refs` somewhere, leading to a usability footgun (stale data, memory leaks).

Another option is to implement `this.refs` as a `Proxy` with dynamic getters that lazily look up the given ref on-demand (maybe with some internal caching). This might be the best perf-wise, but it's complex to implement.

Whatever we do, I would prefer to optimize for the 99.9% usecase. E.g. `Object.keys()` or `Reflect.ownKeys()` should probably return `[]` or throw an error, because I can't imagine any use case that wouldn't be better served by `querySelectorAll`.

Because I'm a perf guy, I'm currently leaning towards `this.refs === this.refs`, and `refs` is always the same Object (a Proxy) for the entire lifespan of the component.

> Shall all the know refs in the template be set to `undefined` in advance?

Yes. A common pattern developers will use is something like `if (this.refs.foo !== undefined) { /* ... */ }`, to check if the node exists yet or not. We already see this pattern in the wild with `querySelector` and `firstChild`.

> When an element with a ref is removed, is the ref set to `undefined`, or is the property removed from the refs object?

Depends on whether we go with the Proxy approach or not.  If `Object.keys()` or `Reflect.ownKeys()` returns empty / throws an error, then it won't matter.

Reply:
I'm ok with the Proxy, but then we have to talk about compat mode, bummer!

I don't think returning empty keys is going to work well... what if the user does `Object.freeze(this.ref)`? what kind of problems do you see here?

Reply:
By the time this feature ships, will we still be supporting IE11? Can we just make the feature not work in IE11?

> what if the user does `Object.freeze(this.ref)` ?

If we can't throw an error on `Object.freeze()` (I don't know if this is possible), then I would just call it undefined behavior. Users should not be doing this.

Reply:
I think we should be careful about this, maybe we should chat offline.

Reply:
I am also sharing the same concerns regarding introducing a new Proxy object.

An alternative approach would create a ref object with getters. Here is a small example of what I have in mind: [link](https://jsfiddle.net/emx1r9v7/1/).

Reply:
Conceptually I'm fine with this approach, I'm just a bit worried about the potential overhead of creating a brand-new `refs` object every time anything changes.

Maybe we can be smart and key off of the list of ref names (array of strings)? This should mitigate the over-creation. Or we can be even smarter and generate the object in the compiler. In either case I would be fine with it.